validate postId in the site-editor.php

mi5t4n opened the pull request:
Added validation for the `postId` parameter in `site-editor.php` to ensure that the post exists before proceeding. If the post ID is invalid or the post does not exist, the script will now terminate with an error message, preventing potential errors or unintended behavior.

<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/61796

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

mi5t4n committed:
validate postId in the site-editor.php

Added validation for the `postId` parameter in `site-editor.php` to ensure that the post exists before proceeding. If the post ID is invalid or the post does not exist, the script will now terminate with an error message, preventing potential errors or unintended behavior.

peterwilsoncc commented on src/wp-admin/site-editor.php:
```diff
@@ -80,6 +80,44 @@ static function ( $classes ) {
+			}
+			break;
+	}
+}
+
+
```
CS: Remove double line break.

Edit: You'll need to do this manually as the inline suggestion doesn't appear to like white space changes.

peterwilsoncc commented on src/wp-admin/site-editor.php:
```diff
@@ -80,6 +80,44 @@ static function ( $classes ) {
 	}
 }
 
+if ( isset( $_GET['postId'] ) ) {
```
Avoids `Warning: Undefined array key "postType"`

```suggestion
if ( isset( $_GET['postId'] ) && isset( $_GET['postType'] ) ) {
```

peterwilsoncc commented on src/wp-admin/site-editor.php:
```diff
@@ -80,6 +80,44 @@ static function ( $classes ) {
+		case 'page':
+			$post = get_post( (int) $_GET['postId'] );
+
+			if ( null === $post ) {
+				wp_die( __( 'Invalid post ID.' ) );
+			}
```
```suggestion
			if ( null === $post || 'page' !== get_post_type( $post ) ) {
				wp_die( __( 'Invalid page ID.' ) );
			}
```

peterwilsoncc commented on src/wp-admin/site-editor.php:
```diff
@@ -80,6 +80,44 @@ static function ( $classes ) {
+		case 'wp_block':
+			$post = get_post( (int) $_GET['postId'] );
+
+			if ( null === $post ) {
```
```suggestion
			if ( null === $post || 'wp_block' !== get_post_type( $post ) ) {
```

peterwilsoncc reviewed (commented):
I've added a couple of notes inline.

While testing I noticed that going to `wp-admin/site-editor.php?postType=page&canvas=edit` ignores the `postType` parameter and loads the `Blog Home` template.

It suggests that further validation of the user input via the URL is required

dream-encode commented:
Though it's probably beyond the scope of the original ticket, I see a lot of duplication of `isset( $_GET['postType'] )` and `sanitize_key( $_GET['postType'] )`.

Wondering if we could have something like a default `$current_post_type = false|null` then check and sanitize it once for reuse in the rest of the file.  Same could pro be done for `$_GET['postId']` and `$_GET['path']`?

Could probably move the early bails a bit higher in the file as well, to save some unnecessary code running when it's not needed?  Just based off a brief look, unless I am missing something, I don't see anything between lines 76 and 118 that is in any way dependent on any of the code running above.

benniledl commented:
would be nice to have an error message for various post statuses as in #7215

mi5t4n committed:
Merge branch 'WordPress:trunk' into fix/61796

mi5t4n committed:
Ensure the post type is not wp_block

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

mi5t4n committed:
Ensure the post type is not page.

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

mi5t4n committed:
Add a check to ensure `postType` exists before proceeding with the request.

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

mi5t4n committed:
Introduced $post_type_param variable with sanitize value of $_GET["postType"].

mi5t4n committed:
Bail early if postId and postType are not valid, so that the unnecessary code does not run.

mi5t4n commented:
@dream-encode  @peterwilsoncc Thank you for the suggestions. I have made the changes.

mi5t4n committed:
removed the new line.

peterwilsoncc commented on src/wp-admin/site-editor.php:
```diff
@@ -19,10 +19,11 @@
 }
 
-$is_template_part        = isset( $_GET['postType'] ) && 'wp_template_part' === sanitize_key( $_GET['postType'] );
+$post_type_param         = sanitize_key( $_GET['postType'] );
```
```suggestion
$post_type_param         = isset( $_GET['postType'] ) ? sanitize_key( wp_unslash( $_GET['postType'] ) ) : '';
```

peterwilsoncc commented on src/wp-admin/site-editor.php:
```diff
@@ -19,10 +19,11 @@
 
-$is_template_part        = isset( $_GET['postType'] ) && 'wp_template_part' === sanitize_key( $_GET['postType'] );
+$post_type_param         = sanitize_key( $_GET['postType'] );
+$is_template_part        = isset( $_GET['postType'] ) && 'wp_template_part' === $post_type_param;
```
```suggestion
$is_template_part        = $post_type_param && 'wp_template_part' === $post_type_param;
```

peterwilsoncc commented on src/wp-admin/site-editor.php:
```diff
@@ -19,10 +19,11 @@
 $is_template_part_path   = isset( $_GET['path'] ) && 'wp_template_partall' === sanitize_key( $_GET['path'] );
 $is_template_part_editor = $is_template_part || $is_template_part_path;
-$is_patterns             = isset( $_GET['postType'] ) && 'wp_block' === sanitize_key( $_GET['postType'] );
+$is_patterns             = isset( $_GET['postType'] ) && 'wp_block' === $post_type_param;
```
```suggestion
$is_patterns             = $post_type_param && 'wp_block' === $post_type_param;
```

peterwilsoncc commented on src/wp-admin/site-editor.php:
```diff
@@ -34,6 +35,43 @@
 }
 
+// Validate postId and postType.
+if ( isset( $_GET['postId'] ) && isset( $_GET['postType'] ) ) {
```
```suggestion
if ( isset( $_GET['postId'] ) && $post_type_param ) {
```

peterwilsoncc reviewed (commented):
I've added a few notes inline for additional reuse of `$post_type_param` and to avoid a potential PHP error if the value is not set.

hellofromtonya committed:
Reuse and avoid a PHP error if the value is not set

Additional reuse of `$post_type_param` and to avoid a potential PHP error if the value is not set.

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

hellofromtonya commented:
Before 6.6.0, it redirected to `site-editor.php?path=%2Fpage&canvas=view`. Granted that did not indicate the _why_. 

Why `wp_die()`?

My concern is two-fold:

* Before this bug: It redirected, but stayed in the Site Editor. 
* Will the user understand how to get back to the editor, i.e. get out of this error page?

I'm wondering:

Could a lighter approach might be a better user experience? "lighter" meaning -> retain the previous behavior to keep the user in the Site Editor, while also displaying a message to inform the user why their requested page did not show.

@dream-encode @peterwilsoncc What do you think?

peterwilsoncc commented:
@hellofromtonya How about passing `link_url` and `link_text` to the function call to return to the site editor?

Untested:

```php
wp_die(
   __( 'Invalid page ID.' ),
   '',
   array(
      'link_url' => admin_url( 'site-editor.php' ),
      'link_text' => 'Return to site editor'
   )
);
```

I think hitting these error messages will require some degree of messing around with the URLs so it's safe to assume a certain amount of advanced user knowledge.

costdev commented on src/wp-admin/site-editor.php:
```diff
@@ -34,6 +35,43 @@
+			break;
+
+		case 'wp_template':
+			$block_template = get_block_template( $_GET['postId'] );
```
Noting that `$_GET['postId']` isn't sanitized here or in the next `case`.

mi5t4n replied:
This has been fixed

hellofromtonya commented:
> @hellofromtonya How about passing `link_url` and `link_text` to the function call to return to the site editor?
> 
> Untested:
> 
> ```
> wp_die(
>    __( 'Invalid page ID.' ),
>    '',
>    array(
>       'link_url' => admin_url( 'site-editor.php' ),
>       'link_text' => 'Return to site editor'
>    )
> );
> ```
> 
> I think hitting these error messages will require some degree of messing around with the URLs so it's safe to assume a certain amount of advanced user knowledge.

That works in my testing. It mitigates one of my concerns:
>Will the user understand how to get back to the editor, i.e. get out of this error page?

Let's adding that to each of the new `wp_die()` instances.

hellofromtonya commented on src/wp-admin/site-editor.php:
```diff
@@ -34,6 +35,43 @@
+			$post = get_post( (int) $_GET['postId'] );
+
+			if ( null === $post || 'page' !== get_post_type( $post ) ) {
+				wp_die( __( 'Invalid page ID.' ) );
```
To help with the user's experience, please add the following as [suggested by ](https://github.com/WordPress/wordpress-develop/pull/7188#issuecomment-2395701839)@peterwilsoncc :

```suggestion
				wp_die(
					__( 'Invalid page ID.' ),
					'',
					array(
					   'link_url'  => admin_url( 'site-editor.php' ),
					   'link_text' => 'Return to site editor',
					)
				);
				
```

mi5t4n replied:
Resolved https://github.com/WordPress/wordpress-develop/pull/7188/commits/ff15ab1ecb005dabffb9e9368ba6fc8a2854297d

hellofromtonya commented on src/wp-admin/site-editor.php:
```diff
@@ -34,6 +35,43 @@
+			$post = get_post( (int) $_GET['postId'] );
+
+			if ( null === $post || 'wp_block' !== get_post_type( $post ) ) {
+				wp_die( __( 'Invalid pattern ID.' ) );
```
[Same here](https://github.com/WordPress/wordpress-develop/pull/7188/files#r1792315558), but maybe should link to the Patterns page.

mi5t4n replied:
Resolved https://github.com/WordPress/wordpress-develop/pull/7188/commits/64e135cadceabeb21339c37905bcfb852c4e1cdc

hellofromtonya commented on src/wp-admin/site-editor.php:
```diff
@@ -34,6 +35,43 @@
+			$block_template = get_block_template( $_GET['postId'] );
+
+			if ( null === $block_template ) {
+				wp_die( __( 'Invalid template ID.' ) );
```
[Same here](https://github.com/WordPress/wordpress-develop/pull/7188/files#r1792315558), but maybe should link to the Templates page.

mi5t4n replied:
Resolved https://github.com/WordPress/wordpress-develop/pull/7188/commits/25ecd7533f7517edebbc9627cb474c4570353092

hellofromtonya commented on src/wp-admin/site-editor.php:
```diff
@@ -34,6 +35,43 @@
+			$block_template = get_block_template( $_GET['postId'], 'wp_template_part' );
+
+			if ( null === $block_template ) {
+				wp_die( __( 'Invalid template part ID.' ) );
```
[Same here](https://github.com/WordPress/wordpress-develop/pull/7188/files#r1792315558), but maybe should link to the Templates page.

mi5t4n replied:
Resolved https://github.com/WordPress/wordpress-develop/pull/7188/commits/39669ab911a9ac069ee53ce328713357c6dfb4bb

mi5t4n committed:
Added a link to the site editor in wp_die() when the page does not exist.

mi5t4n committed:
Added a link to the patterns page in wp_die() when the pattern does not exist.

mi5t4n committed:
Added a link to the templates page in wp_die() when the template does not exist.

mi5t4n committed:
Added a link to the template parts page in wp_die() when the template part does not exist.

mi5t4n commented:
@costdev @hellofromtonya Thanks a lot for the feedback. I have implemented the requested changes.

costdev commented:
Hi @mi5t4n, thanks for the ping!

My contributions to WordPress Core are currently on hold. I am therefore unable to perform a follow-up review at this time.

Please drop a comment on the ticket to ask if other contributors are able to perform a follow-up review.
